File: python-version/compress.py
import os
import sys
from PIL import Image

SUPPORTED_EXTENSIONS = {".jpg", ".jpeg", ".png", ".webp", ".tiff", ".bmp"}

def get_base_dir():
    if getattr(sys, "frozen", False):
        return os.path.dirname(sys.executable)
    return os.path.dirname(os.path.abspath(__file__))

def compress_image(file_path, output_path, quality=60):
    img = Image.open(file_path)
    img.save(output_path, optimize=True, quality=quality)

    original_size = os.path.getsize(file_path)
    compressed_size = os.path.getsize(output_path)
    saved = ((original_size - compressed_size) / original_size) * 100

    name = os.path.basename(file_path)
    print(f"{name}: {original_size / 1024:.1f}KB -> {compressed_size / 1024:.1f}KB ({saved:.1f}% saved)")

def main():
    base = get_base_dir()
    raw_dir = os.path.join(base, "raw")
    compressed_dir = os.path.join(base, "compressed")

    os.makedirs(raw_dir, exist_ok=True)
    os.makedirs(compressed_dir, exist_ok=True)

    files = [
        f for f in os.listdir(raw_dir)
        if os.path.splitext(f)[1].lower() in SUPPORTED_EXTENSIONS
    ]

    if not files:
        print("No images found in raw/ folder. Add images there and run again.")
        input("Press Enter to exit...")
        return

    print(f"Compressing {len(files)} image(s)...\n")

    for file in files:
        compress_image(
            os.path.join(raw_dir, file),
            os.path.join(compressed_dir, file),
        )

    print("\nDone! Compressed images are in the compressed/ folder.")
    input("Press Enter to exit...")

if __name__ == "__main__":
    main()

File: compress.js
const fs = require("fs");
const path = require("path");
const sharp = require("sharp");

const RAW_DIR = path.join(__dirname, "raw");
const COMPRESSED_DIR = path.join(__dirname, "compressed");

const SUPPORTED_EXTENSIONS = [".jpg", ".jpeg", ".png", ".webp", ".tiff", ".avif"];

// Ensure directories exist
if (!fs.existsSync(RAW_DIR)) fs.mkdirSync(RAW_DIR);
if (!fs.existsSync(COMPRESSED_DIR)) fs.mkdirSync(COMPRESSED_DIR);

async function compressImage(filePath) {
  const ext = path.extname(filePath).toLowerCase();
  const fileName = path.basename(filePath);
  const outputPath = path.join(COMPRESSED_DIR, fileName);

  let pipeline = sharp(filePath);
  const quality = 60;
  if (ext === ".png") {
    pipeline = pipeline.png({ quality });
  } else if (ext === ".jpg" || ext === ".jpeg") {
    pipeline = pipeline.jpeg({ quality });
  } else if (ext === ".webp") {
    pipeline = pipeline.webp({ quality });
  } else if (ext === ".tiff") {
    pipeline = pipeline.tiff({ quality });
  } else if (ext === ".avif") {
    pipeline = pipeline.avif({ quality });
  }

  await pipeline.toFile(outputPath);

  const originalSize = fs.statSync(filePath).size;
  const compressedSize = fs.statSync(outputPath).size;
  const saved = (((originalSize - compressedSize) / originalSize) * 100).toFixed(1);

  console.log(`${fileName}: ${(originalSize / 1024).toFixed(1)}KB -> ${(compressedSize / 1024).toFixed(1)}KB (${saved}% saved)`);
}

async function main() {
  const files = fs.readdirSync(RAW_DIR).filter((f) => {
    return SUPPORTED_EXTENSIONS.includes(path.extname(f).toLowerCase());
  });

  if (files.length === 0) {
    console.log("No images found in raw/ directory. Add images there and run again.");
    return;
  }

  console.log(`Compressing ${files.length} image(s)...\n`);

  for (const file of files) {
    await compressImage(path.join(RAW_DIR, file));
  }

  console.log("\nDone! Compressed images are in the compressed/ directory.");
}

main().catch(console.error);
